README.rst
===========
btfx-trader
===========


.. image:: https://img.shields.io/pypi/v/btfx_trader.svg
    :target: https://pypi.python.org/pypi/btfx-trader/

.. image:: https://travis-ci.org/sentrip/btfx-trader.svg?branch=master
    :target: https://travis-ci.com/sentrip/btfx-trader/

.. image:: https://readthedocs.org/projects/btfx-trader/badge/?version=latest
    :target: https://btfx-trader.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://codecov.io/gh/sentrip/btfx-trader/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/sentrip/btfx-trader

.. image:: https://pyup.io/repos/github/sentrip/btfx_trader/shield.svg
    :target: https://pyup.io/repos/github/sentrip/btfx-trader/
    :alt: Updates



Simple to use wrappers for Bitfinex's web socket api


* Free software: `MIT License <https://github.com/sentrip/btfx-trader/blob/master/LICENSE>`_
* Documentation: https://btfx-trader.readthedocs.io.


Features
--------

* Access to cryptocurrency data with an API similar to queue.Queue
* Simple trading API that responds to account data

Installation
-------------

To install btfx-trader, run this command in your terminal:

.. code-block:: shell

    pip install btfx-trader

Usage
------

To use public data:

.. code-block:: python

    from btfx_trader import PublicData

    q = PublicData(types=['tickers'], symbols=['BTCUSD'])
    q.connect()

    while True:
        data = q.get('tickers', 'BTCUSD')


To make an order:

.. code-block:: python

    from btfx_trader import Trader

    trader = Trader('YOUR_BITFINEX_KEY', 'YOUR_BITFINEX_SECRET')
    trader.connect()
    # Order 0.01 BTC at $10000 per bitcoin
    order_id = trader.order('BTCUSD', 10000, dollar_amount=100)


To cancel an order:

.. code-block:: python

    # for a single order
    trader.cancel(order_id)
    # for multiple orders
    trader.cancel_all(older_than=10)



HISTORY.rst
=======
History
=======

0.1.6 (2018-10-15)
------------------

* Fix btfxwss dependency update error


0.1.5 (2018-06-04)
------------------

* Fix rendering of docs


0.1.4 (2018-06-04)
------------------

* Updated README


0.1.3 (2018-05-28)
------------------

* Updated codecov config to only run with travis-ci builds.


0.1.2 (2018-05-27)
------------------

* Added codecov to docs.


0.1.1 (2018-05-27)
------------------

* Fixed rendering of readme on pypi.


0.1.0 (2018-05-25)
------------------

* First release on PyPI.


tests/conftest.py
import os, sys
sys.path.append(os.path.abspath('../btfx_trader'))
from collections import defaultdict
from threading import Thread, Event
from time import time, sleep
from queue import Queue
import pytest
import random
import btfxwss.queue_processor
import requests
from btfx_trader import PublicData, Trader


class TestWSS:
    def __init__(self, key=None, secret=None, log_level=None, tr=None):
        self.trader = tr
        self.key = key
        self.secret = secret
        self.log_level = log_level
        ev = Event()
        Thread(target=lambda: (sleep(1e-3), ev.set())).start()
        self.conn = type('', (), {'connected': ev})
        self.queue_processor = btfxwss.queue_processor.QueueProcessor(
            data_q=Queue(), log_level='CRITICAL'
        )

    def start(self):
        pass

    def stop(self):
        pass

    def authenticate(self):
        pass

    def subscribe_to_ticker(self, ticker):
        pass

    def unsubscribe_from_ticker(self, ticker):
        pass

    def subscribe_to_trades(self, trades):
        pass

    def unsubscribe_from_trades(self, trades):
        pass

    def subscribe_to_candles(self, candles, timeframe=None):
        pass

    def unsubscribe_from_candles(self, candles, timeframe=None):
        pass

    def new_order(self, **kwargs):
        self.trader._update('on',
              [kwargs.pop('id', None) or random.randint(1000000, 9999999),
               '', '', kwargs.get('symbol'),
               '', time() * 1000,
               float(kwargs.get('amount')), float(kwargs.get('amount')), '', '',
               '', '', '',
               'ACTIVE', '', '', float(kwargs.get('price')), 0., '']
        )

    def cancel_order(self, **kwargs):
        _id = kwargs.pop('id')
        order = self.trader.orders[_id]
        self.trader._update('oc', [
            _id, '', '', 't' + order['symbol'], '',
            order['timestamp'] * 1000, order['remaining'], order['amount'],
            '', '', '', '', '',  'CANCELLED', '', '',
            order['price'], order['executed_price'], ''
        ])


@pytest.fixture
def patched_get_symbols(monkeypatch):
    got_once = False
    first = b'["btcusd","ethusd","ltcusd","ethltc"]'
    second = b'["btcusd","ethusd","ltcusd","ethltc","sanusd"]'

    def fake_get(*args, **kwargs):
        nonlocal got_once
        if not got_once:
            got_once = True
            d = first
        else:
            d = second
        r = requests.Response()
        r.status_code, r._content = 200, d
        return r

    monkeypatch.setattr('requests.get', fake_get)
    yield


@pytest.fixture
def data():
    t = PublicData(
        types=['tickers', 'trades', ('candles', '1m')],
        symbols=['BTCUSD', 'LTCUSD', 'ETHUSD']
    )
    t.wss = TestWSS('test_key', 'test_secret', log_level='CRITICAL')
    return t


@pytest.fixture
def trader():
    t = Trader('test_key', 'test_secret')
    t.wss = TestWSS('test_key', 'test_secret', log_level='CRITICAL', tr=t)
    return t


@pytest.fixture
def setup_trader():
    def wrapped(t):
        _wallets = [['EXCHANGE', s.lower().replace('usd', ''),
                     0., None, None] for s in t.symbols]
        t._update('ws', [['EXCHANGE', 'usd', 10000., None, None]] + _wallets)
        t._prices = defaultdict(float)
        t._prices.update({
            'BTCUSD': 15000,
            'ETHUSD': 700,
            'LTCUSD': 300,
            'DSHUSD': 1000,
            'XRPUSD': 0.2
        })
    return wrapped


@pytest.fixture
def new_order():
    def wrapper(tr, symbol, price, amount, ts=None, _id=None):
        _id = _id or random.randint(1000000, 9999999)
        tr._update('on',
             [_id, '', '', 't'+symbol, '', (ts or time())*1000, amount, amount,
              '', '', '', '', '',  'ACTIVE', '', '', price, 0., '']
        )
        return _id
    return wrapper


@pytest.fixture
def update_order():
    # This is bad but the alternative is a mock Bitfinex trading platform so...
    def wrapper(tr, _id, price, amount, cancel=False, execute=False, ts=None):
        if execute or cancel:
            cmd = 'oc'
            if execute:
                status = 'EXECUTED @ ($%.2f)' % price
            else:
                status = 'CANCELLED'
        else:
            cmd = 'ou'
            if amount != 0:
                status = 'PARTIALLY EXECUTED'
            else:
                status = 'ACTIVE'

        exc_price = tr._orders[_id]['executed_price']
        if amount != 0:
            if exc_price != 0:
                exc_price = (exc_price + price) / 2

            s = tr._orders[_id]['symbol'].lower().replace('usd', '')
            dollars = tr._wallets['usd'] - amount * price
            tr._update('wu', ['EXCHANGE', 'usd', dollars, None, None])
            tr._update('wu', [
                'EXCHANGE', s, tr._wallets[s] + tr._orders[_id]['executed'] + amount, None, None
            ])

        tr._update(cmd, [_id, '', '', 't' + tr._orders[_id]['symbol'], '',
                         (ts or time()) * 1000,
                         tr._orders[_id]['remaining'] - amount,
                         tr._orders[_id]['amount'],
                         '', '', '', '', '',  status, '', '',
                         price, exc_price, ''])
    return wrapper


tests/test_trade.py
import random
import unittest
from collections import defaultdict
from multiprocessing import Queue
from threading import Thread
from time import sleep, time

from btfx_trader.trade import Trader


class FakeWSS:
    def __init__(self, key=None, secret=None, log_level=None):
        self.key = key
        self.secret = secret
        self.log_level = log_level

        self.conn = type('', (), {'connected': type('', (), {'is_set': lambda: True})})
        self.queue_processor = type('', (), {'tickers': defaultdict(Queue)})

        self.orders = Queue()
        self.orders_new = Queue()
        self.orders_update = Queue()
        self.orders_cancel = Queue()
        self.wallets = Queue()

    def start(self):
        pass

    def stop(self):
        pass

    def authenticate(self):
        pass

    def subscribe_to_ticker(self, ticker):
        pass

    def unsubscribe_from_ticker(self, ticker):
        pass

    def new_order(self, **kwargs):
        self.orders_new.put(
            [('on',
              [random.randint(1000000, 9999999), '', '',  kwargs.get('symbol'), '', time() * 1000,
               float(kwargs.get('amount')), float(kwargs.get('amount')),  '', '', '', '', '',
               'ACTIVE', '', '', float(kwargs.get('price')), 0., '']), 0]
        )


class FakeTrader(Trader):
    _wss_client = FakeWSS


class TestTraderWSSUpdates(unittest.TestCase):
    def setUp(self):
        self.trader = FakeTrader('key', 'secret', symbols=['BTCUSD'])

    def tearDown(self):
        self.trader.shutdown()

    def test_init(self):
        pass  # Tests initialization of fake api

    def test_wallets_updates(self):
        # wallet init
        data = [('ws', [
            ['EXCHANGE', 'usd', 100., None, None],
            ['EXCHANGE', 'btc', 10., None, None],
            ['EXCHANGE', 'eth', 5., None, None]
        ]), 0]
        self.trader.wss.wallets.put(data)
        sleep(0.002)  # wait for thread to update
        self.assertEqual(100., self.trader.wallets['usd'], 'Did not update usd in wallet')
        self.assertEqual(10., self.trader.wallets['btc'], 'Did not update btc in wallet')
        self.assertEqual(5., self.trader.wallets['eth'], 'Did not update eth in wallet')
        # wallet updates
        self.trader.wss.wallets.put([('wu', ['EXCHANGE', 'usd', 50., None, None]), 0])
        self.trader.wss.wallets.put([('wu', ['EXCHANGE', 'btc', 20., None, None]), 0])
        sleep(0.004)  # wait for thread to update
        self.assertEqual(50., self.trader.wallets['usd'], 'Did not update usd in wallet')
        self.assertEqual(20., self.trader.wallets['btc'], 'Did not update btc in wallet')
        self.assertEqual(5., self.trader.wallets['eth'], 'Did not update eth in wallet')

    def test_orders_init_execute(self):
        # orders init
        ts = time() * 1000
        data = [('os', [
            [123, '', '', 'tBTCUSD', '', ts, 5., 10., '', '', '', '', '',  'PARTIALLY EXECUTED', '', '', 50., 45., '']
        ]), 0]
        expected_json_init = dict(
                symbol='BTCUSD', price=50., executed_price=45., amount=10.,
                executed=5., remaining=5., status='PARTIALLY', timestamp=ts / 1000,
            )
        self.trader.wss.orders.put(data)
        sleep(0.002)  # wait for thread to update
        self.assertEqual(expected_json_init, self.trader.orders[123], 'Incorrect order on init')
        self.assertEqual(list(self.trader.orders.values())[-1], self.trader.orders[123], 'Incorrect latest order')
        self.assertEqual(1, len(self.trader.orders), 'Incorrect length of orders')
        # execute order
        ts = time() * 1000
        data = [('oc',
                 [123, '', '', 'tBTCUSD', '', ts, 0., 10., '', '', '', '', '',  'EXECUTED @', '', '', 50., 43., '']), 0]
        expected_json_execute = dict(
                symbol='BTCUSD', price=50., executed_price=43., amount=10.,
                executed=10., remaining=0., status='EXECUTED', timestamp=ts / 1000,
            )
        self.trader.wss.orders.put(data)
        sleep(0.002)  # wait for thread to update
        assert 123 not in self.trader.orders, 'Did not remove executed order'
        self.assertEqual(0, len(self.trader.orders), 'Incorrect length of orders')
        self.assertEqual((123, expected_json_execute), list(self.trader.executed_orders.items())[-1],
                         'Incorrect order json added to executed orders')

    def test_orders_updates(self):
        # new order
        ts = time() * 1000
        data = [('on',
                 [321, '', '', 'tBTCUSD', '', ts, 10., 10., '', '', '', '', '',  'ACTIVE', '', '', 50., 0., '']), 0]
        expected_json_new = dict(
                symbol='BTCUSD', price=50., executed_price=0., amount=10.,
                executed=0., remaining=10., status='ACTIVE', timestamp=ts / 1000,
            )
        self.trader.wss.orders.put(data)
        sleep(0.002)  # wait for thread to update
        self.assertEqual(expected_json_new, self.trader.orders[321], 'Incorrect new order')
        self.assertEqual(list(self.trader.orders.values())[-1], self.trader.orders[321], 'Incorrect latest order')
        self.assertEqual(1, len(self.trader.orders), 'Incorrect length of orders')
        # update order
        ts = time() * 1000
        data = [('ou',
                 [321, '', '', 'tBTCUSD', '', ts, 10., 10., '', '', '', '', '',  'ACTIVE', '', '', 25., 0., '']), 0]
        expected_json_update = dict(
                symbol='BTCUSD', price=25., executed_price=0., amount=10.,
                executed=0., remaining=10., status='ACTIVE', timestamp=ts / 1000,
            )
        self.trader.wss.orders.put(data)
        sleep(0.002)  # wait for thread to update
        self.assertEqual(expected_json_update, self.trader.orders[321], 'Incorrect update order')
        self.assertEqual(list(self.trader.orders.values())[-1], self.trader.orders[321], 'Incorrect latest order')
        self.assertEqual(1, len(self.trader.orders), 'Incorrect length of orders')
        # cancel order
        ts = time() * 1000
        data = [('oc',
                 [321, '', '', 'tBTCUSD', '', ts, 10., 10., '', '', '', '', '',  'CANCELLED', '', '', 25., 0., '']), 0]
        self.trader.wss.orders.put(data)
        sleep(0.002)  # wait for thread to update
        assert 321 not in self.trader.orders, 'Did not remove cancelled order'
        self.assertEqual(0, len(self.trader.orders), 'Incorrect length of orders')


class TestTrader(unittest.TestCase):
    def setUp(self):
        self.trader = FakeTrader('key', 'secret', symbols=['BTCUSD', 'ETHUSD', 'LTCUSD', 'DSHUSD', "XRPUSD"])
        _wallets = [['EXCHANGE', s.lower().replace('usd', ''), 0., None, None] for s in self.trader.symbols]
        data = [('ws', [['EXCHANGE', 'usd', 10000., None, None]] + _wallets), 0]
        self.trader.wss.wallets.put(data)
        self.trader._prices = defaultdict(float)
        self.trader._prices.update({
            'BTCUSD': 15000,
            'ETHUSD': 700,
            'LTCUSD': 300,
            'DSHUSD': 1000,
            'XRPUSD': 0.2
        })
        sleep(0.005)

    def tearDown(self):
        self.trader.shutdown()

    def _new_order(self, symbol, price, amount):
        _id = random.randint(1000000, 9999999)
        self.trader._orders[_id] = dict(
                symbol=symbol, price=price, executed_price=0., amount=amount,
                executed=0., remaining=amount, status='ACTIVE', timestamp=time() / 1000,
            )
        return _id

    def _update_order(self, _id, price, amount, execute=False):
        if execute:
            status = 'EXECUTED @ ($%.2f)' % price
        else:
            status = 'PARTIALLY EXECUTED'

        self.trader._orders[_id].update(dict(
                executed_price=price, timestamp=time() / 1000, status=status,
                executed=amount, remaining=self.trader._orders[_id]['amount'] - amount
            ))
        if execute:
            self.trader._executed_orders.append((_id, self.trader._orders[_id]))
            del self.trader._orders[_id]

    def test_order_totals(self):
        self._new_order('BTCUSD', 15000, 0.01)
        self.assertEqual(150., self.trader.order_totals['usd'], 'Incorrect total for buying BTCUSD')
        self._new_order('ETHUSD', 700, 0.1)
        self.assertEqual(220., self.trader.order_totals['usd'], 'Incorrect total for buying ETHUSD, BTCUSD')
        _id = self._new_order('BTCUSD', 15000, -0.005)
        self.assertEqual(220., self.trader.order_totals['usd'], 'Incorrect total for selling BTCUSD')
        self.assertEqual(-0.005, self.trader.order_totals['BTCUSD'], 'Incorrect total for selling BTCUSD')
        self._new_order('BTCUSD', 15000, 0.1)
        self._update_order(_id, 15000, -0.0025)
        self.assertEqual(1720., self.trader.order_totals['usd'], 'Incorrect total for updating sell BTCUSD')
        self.assertEqual(-0.0025, self.trader.order_totals['BTCUSD'], 'Incorrect total for updating sell BTCUSD')
        self._update_order(_id, 15000, -0.005, execute=True)
        self.assertEqual(1720., self.trader.order_totals['usd'], 'Incorrect total for executing sell BTCUSD')
        self.assertEqual(0., self.trader.order_totals['BTCUSD'], 'Incorrect total for updating sell BTCUSD')

    def test_wallets(self):
        self.trader._wallets['btc'] = 0.1
        self.trader._wallets['eth'] = 1.
        _id = self._new_order('BTCUSD', 15000, 0.01)
        self.assertEqual(9850., self.trader.wallets['usd'], 'Incorrect wallet value after buy BTCUSD')
        self.assertEqual(0.1, self.trader.wallets['btc'], 'Incorrect wallet value after buy BTCUSD')
        self._update_order(_id, 15000, 0.01, execute=True)
        self.trader._wallets['usd'] -= 150
        self.trader._wallets['btc'] += 0.01
        self.assertEqual(9850., self.trader.wallets['usd'], 'Incorrect wallet value after execute buy BTCUSD')
        self.assertEqual(0.11, self.trader.wallets['btc'], 'Incorrect wallet value after execute buy BTCUSD')

        _id = self._new_order('ETHUSD', 700, -1.)
        self.assertEqual(9850., self.trader.wallets['usd'], 'Incorrect wallet value after sell ETHUSD')
        self.assertEqual(0., self.trader.wallets['eth'], 'Incorrect wallet value after sell ETHUSD')
        self._update_order(_id, 700, -1., execute=True)
        self.trader._wallets['usd'] += 700
        self.trader._wallets['eth'] -= 1
        self.assertEqual(10550., self.trader.wallets['usd'], 'Incorrect wallet value after execute sell BTCUSD')
        self.assertEqual(0., self.trader.wallets['eth'], 'Incorrect wallet value after execute sell ETHUSD')

    def test_orders_value(self):
        self.assertEqual(0., self.trader.orders_value, 'Incorrect total value with no orders')
        self._new_order('BTCUSD', 15000, 0.01)
        self.assertEqual(150., self.trader.orders_value, 'Incorrect total value for buying BTCUSD')
        self._new_order('ETHUSD', 700, 0.1)
        self.assertEqual(220., self.trader.orders_value, 'Incorrect total value for buying ETHUSD, BTCUSD')
        _id = self._new_order('BTCUSD', 15000, -0.01)
        self.assertEqual(370., self.trader.orders_value, 'Incorrect total value for selling BTCUSD')
        self._update_order(_id, 15150, -0.01, execute=True)
        self.assertEqual(220., self.trader.orders_value, 'Incorrect total value for executing sell BTCUSD')

    def test_value(self):
        self.assertEqual(10000., self.trader.value, 'Incorrect initial value')
        _id = self._new_order('BTCUSD', 15000, 0.01)
        self.assertEqual(10000., self.trader.value, 'Incorrect value during trade')
        self._update_order(_id, 15000, 0.01, execute=True)
        self.trader._wallets['usd'] -= 150
        self.trader._wallets['btc'] += 0.01
        self.assertEqual(10000., self.trader.value, 'Incorrect value after neutral trade')
        self.trader._prices['BTCUSD'] = 30000
        self.assertEqual(10150., self.trader.value, 'Incorrect value after price increase')

    def test_wait_execution(self):
        _id = self._new_order('BTCUSD', 15000, 0.01)
        with self.assertRaises(TimeoutError):
            self.trader.wait_execution(_id, seconds=0.001)
        del self.trader._orders[_id]
        _id = self._new_order('BTCUSD', 15000, 0.01)
        result = []
        t = Thread(target=lambda _idd, _wait: result.append(self.trader.wait_execution(_idd, seconds=_wait)),
                   args=(_id, 0.1))
        t.start()
        self._update_order(_id, 15150, 0.01, execute=True)
        t.join()
        self.assertEqual(1, len(self.trader.executed_orders))
        self.assertEqual((_id, result[0]), list(self.trader.executed_orders.items())[-1])

    def test_position(self):
        self.assertEqual(-1., self.trader.position('BTCUSD'), 'Incorrect position with no coin')
        self.trader._wallets['usd'] -= 5000.
        self.trader._wallets['btc'] += 1 / 3
        self.assertEqual(0., self.trader.position('BTCUSD'), 'Incorrect position with balanced dollars and coin')
        self.trader._wallets['usd'] -= 5000.
        self.trader._wallets['btc'] += 1 / 3
        self.assertEqual(1., self.trader.position('BTCUSD'), 'Incorrect position with no dollars')
        self.trader._wallets['usd'] += 2500.
        self.trader._wallets['btc'] -= 1 / 6
        self.assertEqual(0.5, self.trader.position('BTCUSD'), 'Incorrect position with imbalanced dollars and coin')
        self.trader._wallets['usd'] += 5000.
        self.trader._wallets['btc'] -= 1 / 3
        self.assertEqual(-0.5, self.trader.position('BTCUSD'), 'Incorrect position with imbalanced dollars and coin')

    def test_order_incorrect_params(self):
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 1.)
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 1., ratio=1, dollar_amount=1)
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 0.)
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 15000., ratio=1.5)
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 15000., dollar_amount=15000)  # balance is 10000
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 15000., dollar_amount=20)  # min is 35
        with self.assertRaises(AssertionError):
            self.trader.order('BTCUSD', 15000., dollar_amount=20, trade_type='ASDFDSF')

    def test_order_correct_params(self):
        self.trader._wallets['usd'] = 15000
        self.trader._wallets['btc'] = 1.
        self.trader._wallets['xrp'] = 75000.
        _id = self.trader.order('BTCUSD', 15000., ratio=1)
        self.assertEqual(1., self.trader.orders[_id]['amount'], 'Incorrect order amount for ratio(+)')
        _id = self.trader.order('BTCUSD', 15000., ratio=-1)
        self.assertEqual(-1., self.trader.orders[_id]['amount'], 'Incorrect order amount for ratio(-)')
        self.trader._orders.clear()
        _id = self.trader.order('BTCUSD', 15000., dollar_amount=150)
        self.assertEqual(0.01, self.trader.orders[_id]['amount'], 'Incorrect order amount for dollar amount(+)')
        _id = self.trader.order('BTCUSD', 15000., dollar_amount=-150)
        self.assertEqual(-0.01, self.trader.orders[_id]['amount'], 'Incorrect order amount for dollar amount(-)')
        self.trader._orders.clear()
        _id = self.trader.order('BTCUSD', 15000 / 1.01, ratio=1, pad_price=0.01)
        self.assertEqual(1, self.trader.orders[_id]['amount'], 'Incorrect order amount for pad price(+)')
        _id = self.trader.order('BTCUSD', 15000 / 1.01, ratio=-1, pad_price=0.01)
        self.assertEqual(-1, self.trader.orders[_id]['amount'], 'Incorrect order amount for pad price(-)')
        self.trader._orders.clear()
        _id = self.trader.order('XRPUSD', 0.19, ratio=1, pad_price=0.001)
        self.assertEqual(75000., self.trader.orders[_id]['amount'], 'Incorrect order amount for min pad price(+)')
        _id = self.trader.order('XRPUSD', 0.21, ratio=-1, pad_price=0.001)
        self.assertEqual(-75000., self.trader.orders[_id]['amount'], 'Incorrect order amount for min pad price(-)')


btfx_trader/trade.py
import logging
from collections import defaultdict, OrderedDict, deque
from contextlib import suppress
from queue import Empty
from time import time, sleep
from threading import Thread

import requests
from btfxwss import BtfxWss

log = logging.getLogger(__name__)


def get_symbols():
    return list(
        map(lambda q: q.upper(),
            filter(lambda i: 'usd' in i,
                   requests.get('https://api.bitfinex.com/v1/symbols').json()))
    )


class Trader(Thread):
    _wss_client = BtfxWss
    _min_order_value = 35.
    _max_order_history = 100
    _sleep_time = 0.001
    _trade_types = {'market', 'limit'}

    def __init__(self, key, secret, symbols=None):
        super(Trader, self).__init__()
        if symbols:
            self.symbols = [s.upper() + ('' if 'usd' in s.lower() else 'USD') for s in symbols]
        else:
            self.symbols = get_symbols()

        self.connected = False
        self._prices = defaultdict(float)
        self._wallets = defaultdict(float)
        self._orders = OrderedDict()
        self._executed_orders = deque(maxlen=self._max_order_history)

        self._source_handlers = OrderedDict((
            ('orders', self._handle_order),
            ('orders_new', self._handle_order),
            ('orders_update', self._handle_order),
            ('orders_cancel', self._handle_order),
            ('wallets', self._handle_wallet)
        ))

        self.wss = self._wss_client(key=key, secret=secret, log_level='CRITICAL')
        self.start()

    def _connect(self):
        self.wss.start()
        log.debug('Waiting for btfxwss api socket connection...')
        while not self.wss.conn.connected.is_set():
            sleep(0.1)
        log.debug('Waiting for btfxwss api authentication...')
        self.wss.authenticate()
        log.debug('Subscribing to symbols...')
        for symbol in self.symbols:
            self.wss.subscribe_to_ticker(symbol)
        self.connected = True
        log.info('Successfully initialized')

    @staticmethod
    def _create_order_json(data):
        orders = []
        for _id, _, _, symbol, _, ts, remain_amount, amount, _, _, _, _, _, stat, _, _, price, exec_price, *_ in data:
            orders.append((_id, dict(
                symbol=symbol[1:].upper(),
                price=price,
                executed_price=exec_price,
                amount=amount,
                executed=amount - remain_amount,
                remaining=remain_amount,
                status=stat.split()[0],
                timestamp=ts / 1000,
            )))
        return orders

    @staticmethod
    def _log_order(_id, symbol, action, amount, price):
        log.info('%d (%6s) %-7s %-4s, %.8f for $%.2f at $%.2f', _id, symbol, action,
                 'BUY' if amount > 0 else 'SELL', abs(amount), abs(amount * price), price)

    def _handle_order(self, _type, data):
        if _type == 'os':
            orders = self._create_order_json(data)
            for _id, order in sorted(orders, key=lambda t: t[1]['timestamp']):
                self._orders[_id] = order
        else:
            _id, order = self._create_order_json([data])[0]
            if _type == 'oc':
                if order['status'].startswith('CANCEL'):
                    self._log_order(_id, order['symbol'], 'CANCEL', order['amount'], order['price'])
                else:
                    self._log_order(_id, order['symbol'], 'EXECUTE', order['executed'], order['executed_price'])
                    self._executed_orders.append((_id, order))
                if _id in self._orders:
                    del self._orders[_id]

            elif _type in {'on', 'ou'}:
                self._orders[_id] = order
                self._log_order(_id, order['symbol'], 'SUBMIT', order['amount'], order['price'])

    def _handle_wallet(self, _type, data):
        if _type == 'ws':
            for wallet_type, currency, balance, _, _ in data:
                if wallet_type.lower() == 'exchange':
                    self._wallets[currency.lower()] = float(balance)
        elif _type == 'wu':
            self._wallets[data[1].lower()] = float(data[2])

    def run(self):
        self._connect()
        while self.connected:
            start_loop = time()
            for source, handler in self._source_handlers.items():
                with suppress(Empty):
                    handler(*(getattr(self.wss, source).get_nowait()[0]))
            for symbol in self.symbols:
                with suppress(Empty):
                    data = self.wss.queue_processor.tickers[('ticker', symbol)].get_nowait()[0]
                    self._prices[symbol] = data[0][6]
            sleep(max(0, self._sleep_time - (time() - start_loop)))

    @property
    def prices(self):
        return self._prices.copy()

    @property
    def orders(self):
        return self._orders.copy()

    @property
    def executed_orders(self):
        return OrderedDict(self._executed_orders)

    @property
    def order_totals(self):
        totals = defaultdict(float)
        totals['usd'] = 0.
        for order in self.orders.values():
            if order['remaining'] < 0:
                totals[order['symbol']] += order['remaining']
            else:
                totals['usd'] += order['remaining'] * order['price']
        return totals

    @property
    def wallets(self):
        wallets = self._wallets.copy()
        for symbol, total in self.order_totals.items():
            wallets[symbol.replace('USD', '').lower()] -= abs(total)
        return wallets

    @property
    def orders_value(self):
        total = 0.
        for symbol, t in self.order_totals.items():
            total += abs(t) * (1 if t > 0 else self._prices[symbol])
        return round(total, 2)

    @property
    def value(self):
        total = 0.
        for c, amount in self._wallets.items():
            if c != 'usd':
                amount *= self._prices[c.upper() + 'USD']
            total += amount
        return round(total, 2)

    def cancel(self, _id):
        self.wss.cancel_order(multi=False, id=_id)

    def cancel_all(self, older_than=0):
        now = time()
        for _id, order in self._orders.items():
            if now - order['timestamp'] > older_than:
                self.cancel(_id)

    def order(self, symbol, price, *,
              ratio=None, dollar_amount=None,
              pad_price=None, trade_type='limit', return_id=True):

        assert ratio or dollar_amount, 'Must provide either `ratio` or `dollar_amount`'
        assert not (ratio and dollar_amount), 'Cannot provide both `ratio` and `dollar_amount`'
        assert trade_type.lower() in self._trade_types, \
            'Unknown trade type, try one of these: %s' % str(self._trade_types)[1:-1]

        buying = (ratio or dollar_amount) > 0
        symbol = symbol.upper() + ('USD' if 'usd' not in symbol.lower() else '')
        assert symbol in self.symbols, 'Trader not initialized with %s in symbols'

        if price == 'market' or trade_type == 'market':
            trade_type = 'market'
            price = self._prices[symbol]
            pad_price = 0.001

        if pad_price:
            delta = price * pad_price
            price += max(delta, 0.01) if buying else min(-delta, -0.01)

        assert price >= 0.01, 'Price cannot be less than $0.01'
        max_amount = self.wallets['usd'] / price if buying else self.wallets[symbol.lower().replace('usd', '')]
        if ratio:
            amount = round(max_amount * ratio, 8)
        else:
            amount = round(dollar_amount / price, 8)

        assertion_msg = 'Trade value (${:.2f}) is %s available trade value ($%.2f)'.format(abs(amount))
        assert abs(amount) >= self._min_order_value / price, assertion_msg % ('below minimum', self._min_order_value)
        assert abs(amount) <= max_amount, assertion_msg % ('above maximum', max_amount * price)

        current_order_ids = set(self._orders.keys())
        last_executed_id = (self._executed_orders or [[None, None]])[-1][0]
        self.wss.new_order(
            cid=int(time()),
            type="EXCHANGE %s" % trade_type.upper(),
            symbol="t%s" % symbol,
            amount="%.8f" % amount,
            price="%.4f" % price,
        )

        while return_id:
            for _id in self._orders:
                if _id not in current_order_ids:
                    return _id
            else:
                if len(self._executed_orders) > 0 and self._executed_orders[-1][0] != last_executed_id:
                    return self._executed_orders[-1][0]
                else:
                    sleep(self._sleep_time)

    def wait_execution(self, _id, seconds=1e9):
        start_time = time()
        while time() - start_time < seconds:
            try:
                ids, trades = tuple(zip(*self._executed_orders))
                return trades[ids.index(_id)]
            except ValueError:
                sleep(self._sleep_time)
        raise TimeoutError('Waiting for execution of order %d timed out after %d seconds' % (_id, seconds))

    def position(self, symbol):
        return (self._wallets[symbol.lower().replace('usd', '')] * self._prices[symbol]
                - self._wallets['usd']) / self.value

    def reset_position(self, symbol):
        try:
            _id = self.order(symbol, 'market', dollar_amount=self._wallets['usd'] - self.value / 2)
            self.wait_execution(_id)
        except AssertionError:
            log.error('Position already reset (%.4f)', self.position(symbol))

    def shutdown(self):
        self.connected = False
        for symbol in self.symbols:
            self.wss.unsubscribe_from_ticker(symbol)
        self.wss.stop()
        self.join()
